src/nlp.jl

```diff
@@ -1025,47 +1025,69 @@ function MOI.constraint_expr(d::NLPEvaluator,i::Integer)
-#     for nlexpr in nldata.nlexpr
-#         push!(subexpr, nlexpr.nd)
-#     end
-#
-#     this_subexpr = nldata.nlexpr[x.index]
-#
-#     max_len = length(this_subexpr.nd)
-#
-#     subexpression_order, individual_order = order_subexpressions(Vector{NodeData}[this_subexpr.nd],subexpr)
-#
-#     subexpr_values = Array{Float64}(undef,length(subexpr))
-#
-#     for k in subexpression_order
-#         max_len = max(max_len, length(nldata.nlexpr[k].nd))
-#     end
-#
-#     forward_storage = Array{Float64}(undef,max_len)
-#     partials_storage = Array{Float64}(undef,max_len)
-#     user_input_buffer = zeros(nldata.largest_user_input_dimension)
-#     user_output_buffer = zeros(nldata.largest_user_input_dimension)
-#
-#     for k in subexpression_order # compute value of dependent subexpressions
-#         ex = nldata.nlexpr[k]
-#         adj = adjmat(ex.nd)
-#         subexpr_values[k] = forward_eval(forward_storage,partials_storage,ex.nd,adj,ex.const_values,nldata.nlparamvalues,m.colVal,subexpr_values,user_input_buffer,user_output_buffer)
-#     end
-#
-#     adj = adjmat(this_subexpr.nd)
-#
-#     return forward_eval(forward_storage,partials_storage,this_subexpr.nd,adj,this_subexpr.const_values,nldata.nlparamvalues,m.colVal,subexpr_values,user_input_buffer,user_output_buffer)
-# end
+# NOTE: This is a slow approach that does *a lot* of setup work on each call.
+# See https://github.com/JuliaOpt/JuMP.jl/issues/746.
+"""
+    value(ex::NonlinearExpression, var_value::Function)
+
+Evaluate `ex` using `var_value(v)` as the value for each variable `v`.
+"""
+function value(ex::NonlinearExpression, var_value::Function)
+    model = ex.m
+
+    nlp_data::NLPData = model.nlp_data
+
+    moi_index_to_consecutive_index = Dict{MOI.VariableIndex, Int}()
+    variable_indices = MOI.get(model, MOI.ListOfVariableIndices())
+    variable_values = Array{Float64}(undef, length(variable_indices))
+    for (consecutive_index, moi_index) in enumerate(variable_indices)
+        moi_index_to_consecutive_index[moi_index] = consecutive_index
+        jump_var = VariableRef(model, moi_index)
+        variable_values[consecutive_index] = var_value(jump_var)
+    end
+
+    subexpressions = Array{Vector{NodeData}}(undef, 0)
+    for nl_expr in nlp_data.nlexpr
+        push!(subexpressions,
+              replace_moi_variables(nl_expr.nd,
+                                    moi_index_to_consecutive_index))
+    end
+
+    original_ex = nlp_data.nlexpr[ex.index]
+    ex_nd = replace_moi_variables(original_ex.nd,
+                                  moi_index_to_consecutive_index)
+    main_expressions = Vector{NodeData}[ex_nd]
+    subexpression_order, _ = order_subexpressions(main_expressions,
+                                                  subexpressions)
+
+    max_len = length(ex_nd)
+    for k in subexpression_order
+        max_len = max(max_len, length(subexpressions[k]))
+    end
+
+    subexpr_values = Array{Float64}(undef, length(subexpressions))
```

Comment:
Using `Vector` is more readable than `Array` since we know that we are building a vector